src/backend/distributed/executor/partitioned_intermediate_results.c

```diff
@@ -44,57 +44,48 @@ typedef struct PartitionedResultDestReceiver
 	/* public DestReceiver interface */
 	DestReceiver pub;
 
-	/* partition file $i is stored at file named $resultIdPrefix_$i. */
-	char *resultIdPrefix;
+	/* on lazy startup we only startup the DestReceiver once they receive a tuple */
+	bool lazyStartup;
 
-	/* use binary copy or just text copy format? */
-	bool binaryCopy;
-
-	/* used for deciding which partition a shard belongs to. */
-	CitusTableCacheEntry *shardSearchInfo;
-
-	MemoryContext perTupleContext;
-
-	/* how does stream tuples look like? */
-	TupleDesc tupleDescriptor;
+	/* cached startup information for use in lazy startup */
+	struct
+	{
+		int operation;
+		TupleDesc tupleDescriptor;
+	} startupCache;
 
 	/* which column of streamed tuples to use as partition column? */
 	int partitionColumnIndex;
 
 	/* how many partitions do we have? */
 	int partitionCount;
 
-	/*
-	 * Tuples for partition[i] are sent to partitionDestReceivers[i], which
-	 * writes it to a result file.
-	 */
+	/* used for deciding which partition a shard belongs to. */
+	CitusTableCacheEntry *shardSearchInfo;
+
+	/* Tuples matching shardSearchInfo[i] are sent to partitionDestReceivers[i] */
 	DestReceiver **partitionDestReceivers;
+	DestReceiver **partitionDestReceiversStarted;
```

Comment:
could this be a `bool *`?

Reply:
Short answer is yes :)

Longer answer, if you look in the history I actually started with such an approach, albeit not a `bool *` but a `Bitmapset`. With the idea to save on memory for the bookkeeping here, which is probably also your reasoning for the `bool *`

Although not measured, I went for 2 `DestReceiver **`'s instead to directly have the pointer it would need to dispatch the tuple to in the most common case. Once all receivers are started it would solely hit the `partitionDestReceiversStarted` array. Only on a miss it would need to look in a separate array to find the Receiver and start it.

What would the reasoning/benefit be of having it in a `bool *`?

Reply:
> What would the reasoning/benefit be of having it in a bool *?

Just to make it easier to understand. There's only 1 set of DestReceiver, while the code suggests there are 2. 

Single pointer dereference per tuple seems irrelevant in terms of performance / premature optimization.

Reply:
`bool *` easier to understand compared to a Bitmapset I assume.

Reply:
I haven't thought about what Bitmapset would look like vs. bool*. In any case, I don't think the memory pressure of a byte per DestReceiver is significant either, so whichever looks nicer.

Reply:
startup:
https://github.com/citusdata/citus/pull/5432/commits/23a4ed5eec82cec0c4dfc892b614110088e90db6#diff-99075aacff38be30e147fe56806cce47b6056bcd6bd812cfe0293c18cb2a173dR415-R421

shutdown:
https://github.com/citusdata/citus/pull/5432/commits/07408d9ff12fad09a8ee6cee9b447ffbebbd7212#diff-99075aacff38be30e147fe56806cce47b6056bcd6bd812cfe0293c18cb2a173dR505-R509

especially iterating over started ones looks slightly more compact.

Reply:
I guess I'm fine with either approach